File: array_problems/sortArray/sortArray.py
class ArrayProblems:
    def __init__(self, arr):
        self.arr = arr.copy()
        
    def Quick(self, arr=None):
        if arr is None:
            arr = self.arr.copy()
        if len(arr) < 2:
            return arr
        pivot = arr[0]
        smaller = [item for item in arr[1:] if item < pivot]
        bigger = [item for item in arr[1:] if item > pivot]
        equal = [item for item in arr if item == pivot]
        return self.quick(smaller) + equal + self.quick(bigger)

    def Bubble(self):
        arr = self.arr.copy()
        length = len(arr)
        for passover in range(length):
            for index in range(length - 1):
                if arr[index] > arr[index + 1]:
                    arr[index], arr[index + 1] = arr[index + 1], arr[index]
        return arr

    def Select(self):
        arr = self.arr.copy()
        length = len(arr)
        for passes in range(length):
            minimum = passes
            for i in range(passes, length):
                if arr[i] < arr[minimum]:
                    minimum = i
            if minimum != passes:
                arr[passes], arr[minimum] = arr[minimum], arr[passes]
        return arr

    def Insert(self):
        arr = self.arr.copy()
        length = len(arr)
        for i in range(1, length):
            index = i - 1
            temporary = arr[i]
            while index >= 0 and arr[index] > temporary:
                arr[index + 1] = arr[index]
                index -= 1
            arr[index + 1] = temporary
        return arr

    def Simple(self):
        arr = self.arr.copy()
        length = len(arr)
        for i in range(length):
            for j in range(length):
                if arr[i] < arr[j]:
                    arr[i], arr[j] = arr[j], arr[i]
        return arr

    def Merge_sort(self, arr=None):
        if arr is None:
            arr = self.arr.copy()
        if len(arr) <= 1:
            return arr
        middle = len(arr) // 2
        left = self.Merge_sort(arr[:middle])
        right = self.Merge_sort(arr[middle:])
        return self.Merge(left, right)

    def Merge(self, left, right):
        merged = []
        while len(left) > 0 and len(right) > 0:
            if left[0] <= right[0]:
                merged.append(left[0])
                left = left[1:]
            else:
                merged.append(right[0])
                right = right[1:]
        merged.extend(left)
        merged.extend(right)
        return merged


arr_problems = ArrayProblems([0, 43, 3, 2, 3, 4, 6])

quick_sort = arr_problems.Quick()
select_sort = arr_problems.Select()
insert_sort = arr_problems.Insert()
bubble_sort = arr_problems.Bubble()
simple_sort = arr_problems.Simple()
merge_sort = arr_problems.Merge_sort()

print("quick sort:", quick_sort)
print("select sort:", select_sort)
print("insert sort:", insert_sort)
print("bubble sort:", bubble_sort)
print("simple sort:", simple_sort)
print("merge sort:", merge_sort)

File: array_problems/sortArray/sortArray.go
package main

import "fmt"

type ArrayProblems struct {
    arr []int
}

func NewArrayProblems(arr []int) *ArrayProblems {
    return &ArrayProblems{arr: copyArray(arr)}
}

func (ap *ArrayProblems) Quick(items []int) []int {
	if items == nil {
		items = copyArray(ap.arr)
	}
	if len(items) < 2 {
		return items
	}
	pivot := items[0]
	var smaller []int
	var bigger []int
	var equal []int
	for _, item := range items {
		if item < pivot {
			smaller = append(smaller, item)
		} else if item > pivot {
			bigger = append(bigger, item)
		} else {
			equal = append(equal, item)
		}
	}
	return append(append(ap.Quick(smaller), equal...), ap.Quick(bigger)...)
}

func (ap *ArrayProblems) Bubble(items []int) []int {
	if items == nil {
		items = copyArray(ap.arr)
	}
	length := len(items)
	for passover := 0; passover < length; passover++ {
		for index := 0; index < length-1; index++ {
			if items[index] > items[index+1] {
				items[index], items[index+1] = items[index+1], items[index]
			}
		}
	}
	return items
}

func (ap *ArrayProblems) Select(items []int) []int {
	if items == nil {
		items = copyArray(ap.arr)
	}
	length := len(items)
	for passes := 0; passes < length; passes++ {
			min := passes
			for i := passes; i < length; i++ {
					if items[i] < items[min] {
							min = i
					}
			}
			if min != passes {
					items[passes], items[min] = items[min], items[passes]
			}
	}
	return items
}

func (ap *ArrayProblems) Insert(items []int) []int {
	if items == nil {
			items = copyArray(ap.arr)
	}
	length := len(items)
	for i := 1; i < length; i++ {
			index := i - 1
			temporary := items[i]
			for index >= 0 && items[index] > temporary {
				items[index+1] = items[index]
				index--
			}
			items[index+1] = temporary
	}
	return items
}

func (ap *ArrayProblems) Simple(items []int) []int {
	if items == nil {
		items = copyArray(ap.arr)
	}
	length := len(items)
	for i := 0; i < length; i++ {
		for j := 0; j < length; j++ {
			if items[i] < items[j] {
				items[i], items[j] = items[j], items[i]
			}
		}
	}
	return items
}

func (ap *ArrayProblems) MergeSort(items []int) []int {
	if items == nil {
		items = copyArray(ap.arr)
	}
	if len(items) <= 1 {
		return items
	}
	middle := len(items) / 2
	left := ap.MergeSort(items[:middle])
	right := ap.MergeSort(items[middle:])
	return ap.Merge(left, right)
}

func (ap *ArrayProblems) Merge(left, right []int) []int {
	merged := []int{}
	for len(left) > 0 && len(right) > 0 {
		if left[0] <= right[0] {
			merged = append(merged, left[0])
			left = left[1:]
		} else {
			merged = append(merged, right[0])
			right = right[1:]
		}
	}
	merged = append(merged, left...)
	merged = append(merged, right...)
	return merged
}

func copyArray(arr []int) []int {
	newArr := make([]int, len(arr))
	copy(newArr, arr)
	return newArr
}

func main() {
	arr := []int{0, 43, 3, 2, 3, 4, 6}
	sort := NewArrayProblems(arr)

	quickSort := sort.Quick(nil)
	selectSort := sort.Select(nil)
	insertSort := sort.Insert(nil)
	bubbleSort := sort.Bubble(nil)
	simpleSort := sort.Simple(nil)
	mergeSort := sort.MergeSort(nil)

	fmt.Println("quick sort:", quickSort)
	fmt.Println("select sort:", selectSort)
	fmt.Println("insert sort:", insertSort)
	fmt.Println("bubble sort:", bubbleSort)
	fmt.Println("simple sort:", simpleSort)
	fmt.Println("merge sort:", mergeSort)
}
